src/transformers/models/wav2vec2/modeling_wav2vec2.py

```diff
@@ -1854,16 +1850,15 @@ def forward(
             input_values,
             output_attentions=output_attentions,
             output_hidden_states=output_hidden_states,
-            return_dict=return_dict,
+            return_dict=True,
```

Comment:
Usually we try to avoid forcing `return_dict` in the call to any modules, and leave it to the setting input by the user -> can we respect the setting of `return_dict` used here? i.e. set:
```python
            return_dict=return_dict,
````

Reply:
I didn't change this at first, but faced a lot of problems when I try to access the returned `attention_mask` in a tuple-indexing style. Following existing code for `hidden_states`, where a global variable `_HIDDEN_STATES_START_POSITION = 2` is defined, I tried to define `_ATTENTION_MASK_START_POSITION = 4` and use that for indexing. This becomes a big problem because the tuple output might not include `hidden_states`, thus the index 4 is wrong. The cleaner solution for me was to force `return_dict=True` for the sub-model, thus I can always access `outputs.attention_mask` reliably and cleanly.

The alternative solution is to check if `output_hidden_states` is True (which can be overridden by `self.config.use_weighted_layer_sum`), then use index=4 for `attention_mask`, otherwise use index=3. Let me know if you prefer which solution.

(Also I just noticed `output_hidden_states` argument in `forward()` for derived models (ForMaskedLM, XVector...) does not do anything since it won't affect the final output)

Reply:
What test do you run to show that `return_dict=True` breaks the graph? From my local testing it seems like `return_dict=True` is fine, but graph breaking occurs due to `getattr(self.config, "apply_spec_augment", True)` (I have replaced `hidden_states[mask/indices] = something` with `hidden_states.masked_filled()` to avoid in-place ops)

```python
import torch
from transformers.models.wav2vec2 import Wav2Vec2Model, Wav2Vec2ForSequenceClassification

# model = Wav2Vec2Model.from_pretrained("facebook/wav2vec2-base")
model = Wav2Vec2ForSequenceClassification.from_pretrained("facebook/wav2vec2-base")
model = torch.compile(model, fullgraph=True)

x = torch.randn(2, 16000 * 2)

out = model(x)
```

Stack trace

```python
  File "/Users/thien/miniforge3/envs/pytorch/lib/python3.10/site-packages/torch/_dynamo/variables/base.py", line 225, in call_hasattr
    unimplemented(f"hasattr: {repr(self)}")
  File "/Users/thien/miniforge3/envs/pytorch/lib/python3.10/site-packages/torch/_dynamo/exc.py", line 71, in unimplemented
    raise Unsupported(msg)
torch._dynamo.exc.Unsupported: hasattr: HFPretrainedConfigVariable()

from user code:
   File "/Users/thien/github/transformers/src/transformers/models/wav2vec2/modeling_wav2vec2.py", line 1578, in forward
    hidden_states = self._mask_hidden_states(
  File "/Users/thien/github/transformers/src/transformers/models/wav2vec2/modeling_wav2vec2.py", line 1508, in _mask_hidden_states
    if not getattr(self.config, "apply_spec_augment", True):
```

If I remove `fullgraph=True` in `torch.compile()`, the model compiles without error.

Reply:
Hmm okay maybe `return_dict=True` is okay in this case. Can you compile with `fullgraph=True` if you skip the check  `if not getattr(self.config, "apply_spec_augment", True)`?

More generally, are we moving away from forcing `return_dict` @amyeroberts in the interest of `torch.compile`? Or is it fine in an instance like this where it doesn't seem to affect it?

Reply:
Replacing 

```python
if not getattr(self.config, "apply_spec_augment", True)
```

with

```
if not self.config.apply_spec_augment
```

makes the model compile without any errors (`fullgraph=True`).

Do you have the final decision whether to propagate `return_dict`, or overriding it is fine? This is the only issue left for this PR.

Reply:
I'm afraid we can't do that @gau-nernst, sorry! The config attribute `apply_spec_augment` was not part of the original Wav2Vec2 PR. It was only added later in a follow-up. This means that users running the original version of Wav2Vec2 with an earlier version of `transformers` will not have the config attribute `apply_spec_augment`, so we can't call `config.apply_spec_augment` straight like this, since the config doesn't have it. This is why it's gated with 
`getattr(self.config, "apply_spec_augment", True)` in the modelling code

Reply:
Regarding whether it's ok to force `return_dict=True`, it would be great to get the opinion of a core maintainer! cc @ArthurZucker @amyeroberts

Essentially, we want to include an extra output for the Wav2Vec2 model output. @gau-nernst has highlighted that the most straightforward way of getting this output is by forcing `return_dict=True` in the call to the Wav2Vec2 model: https://github.com/huggingface/transformers/pull/25471#discussion_r1295362009. Is this ok with you if we check that the model can still be compiled this way?

Reply:
@sanchit-gandhi I'm not proposing to replace it to `if not self.config.apply_spec_augment`. Making Wav2Vec2 `torch.compile()`-able is out of the scope of this PR. I just want to show that the problem that prevents Wav2Vec2 from being able to be compiled is due to the `getattr()` function, not due to overriding `return_dict`.

Let me know once your team has reached a decision regarding overiding `return_dict`.

On a separate note, I don't know how HF handles the config system, but it's kinda brittle to not be able to add new attributes to the config class in newer versions without resorting to call `getattr()`. There are definitely ways to handle that better. (Again, this is out of scope of this PR, I'm just commenting).

Reply:
@sanchit-gandhi @gau-nernst Yes, we don't force `return_dict=True` as it messes things up for `torch.compile`, unfortunately. It's true `return_dict=True` makes things more straightforward and less error prone (we can access name attributes rather than indexing) but this is true for pretty much all of our models.

Reply:
Hey @amyeroberts - just wondering if you had a chance to circle back here? As @gau-nernst mentioned in their prior comment, it looks like `torch.compile` is not affected by setting `return_dict=True` in this module: https://github.com/huggingface/transformers/pull/25471#discussion_r1306286367

Is it ok to force this value in this case?

Reply:
Hi @gau-nernst, apologies for the delay. I thought there were tests for `torch.compile` but it seems that there aren't yet (cc @ydshieh). I've had to convert models last month because of setting `return_dict=True`, and this has been a [design pattern for a while](https://github.com/huggingface/transformers/pull/22970#discussion_r1195252303). If you can successfully compile and use this model setting `return_dict=True` then I don't see why we couldn't do this.

Reply:
It's also a case in `diffusers` that setting `return_dict=False` is necessary to enable torch compile, e.g. https://github.com/huggingface/diffusers/pull/4738. But if we've comprehensively shown that torch compile works with `return_dict=True`, I also don't see any issue in allowing this!